hypershift-operator/controllers/scheduler/dedicated_content_serving_nodes.go

```diff
@@ -0,0 +1,204 @@
+			return ctrl.Result{}, fmt.Errorf("failed to delete node: %w", err)
+		}
+		log.Info("Node deleted")
+	} else {
+		log.Info("The hosted cluster exists, will check again later.")
+		return ctrl.Result{RequeueAfter: 1 * time.Minute}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+type DedicatedServingComponentScheduler struct {
+	client.Client
+	createOrUpdate upsert.CreateOrUpdateFN
+}
+
+func (r *DedicatedServingComponentScheduler) SetupWithManager(mgr ctrl.Manager, createOrUpdateProvider upsert.CreateOrUpdateProvider) error {
+
+	r.createOrUpdate = createOrUpdateProvider.CreateOrUpdate
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedCluster{}).
+		WithOptions(controller.Options{
+			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+			MaxConcurrentReconciles: 10,
+		}).Named("DedicatedServingComponentScheduler")
+	return builder.Complete(r)
+}
+
+func (r *DedicatedServingComponentScheduler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	hcluster := &hyperv1.HostedCluster{}
+	log := ctrl.LoggerFrom(ctx, "controller", "DedicatedServingComponentScheduler", "hostedcluster", req.NamespacedName.String())
+	err := r.Get(ctx, req.NamespacedName, hcluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("hostedcluster not found, aborting reconcile", "name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get cluster %q: %w", req.NamespacedName, err)
+	}
+	if !hcluster.DeletionTimestamp.IsZero() {
+		log.Info("hostedcluster is deleted, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	if isolationMode := hcluster.Annotations[hyperv1.IsolationModeAnnotation]; isolationMode != hyperv1.DedicatedRequestServingComponentsIsolationMode {
+		log.Info("hostedcluster does not use isolated request serving components, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	if scheduled := hcluster.Annotations[hyperv1.HostedClusterScheduledAnnotation]; scheduled == "true" {
+		log.Info("hostedcluster is already scheduled, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	nodeList := &corev1.NodeList{}
+	if err := r.List(ctx, nodeList, client.HasLabels{hyperv1.RequestServingComponentLabel}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	nodesToUse := map[string]*corev1.Node{}
+	// first, find any existing nodes already labeled for this hostedcluster
+	for i := range nodeList.Items {
+		node := &nodeList.Items[i]
+		zone, hasZoneLabel := node.Labels["topology.kubernetes.io/zone"]
+		if !hasZoneLabel {
+			continue
+		}
+		hcLabel, hasHCLabel := node.Labels[hyperv1.HostedClusterLabel]
+		if !hasHCLabel {
+			continue
+		}
+		if hcLabel == fmt.Sprintf("%s-%s", hcluster.Namespace, hcluster.Name) {
+			nodesToUse[zone] = node
+			log.Info("Found existing node for hosted cluster", "node", node.Name, "zone", zone)
+		}
+	}
+
+	if len(nodesToUse) < 2 {
```

Comment:
Instead splitting in two loops have we considering one loop, order, filter out?
Is there anything else beyond this controller setting the hcLabel?

Reply:
We need 2 loops. One to determine if there are nodes that were already assigned to the cluster, the second one to find  available nodes we need. I don't think we can condense into a single loop.

Reply:
we could use something like this to construct a list where the first 2 nodes are always the ones we want to use: 
loop over nodes:
  1. if hasHCLabel, check first node in the list `nodesToUse`
      - if it has same zone -> replace it
      - otherwise -> prepend the node to the start of the list
  2. if not hasHCLabel -> free node, check last node in the list `nodesToUse`
      - if it has same zone -> skip
      - otherwise -> append the node to the end of the list

Reply:
with a linkedList it looks like this:
```	
nodesToUse := list.New()
for _, node := range nodes {
	zone, hasZoneLabel := node.Labels["topology.kubernetes.io/zone"]
	if !hasZoneLabel {
		continue
	}
	hcLabel, hasHCLabel := node.Labels["hypershift.io/cluster"]
	if !hasHCLabel {
		back := nodesToUse.Back()
		if back == nil || back.Value.(corev1.Node).Labels["topology.kubernetes.io/zone"] != zone {
			nodesToUse.PushBack(node)
		}
	} else if hcLabel == fmt.Sprintf("%s-%s", hcluster.Namespace, hcluster.Name) {
		if nodesToUse.Front() == nil {
			nodesToUse.PushFront(node)
			continue
		}
		frontZone := nodesToUse.Front().Value.(corev1.Node).Labels["topology.kubernetes.io/zone"]
		if frontZone != zone {
			nodesToUse.PushFront(node)
		} else {
			// replace
			nodesToUse.Front().Value = node
		}
	}
}

// take only first 2 nodes
count := 0
for e := nodesToUse.Front(); e != nil && count < 2; e = e.Next() {
	count++
	node := e.Value.(corev1.Node)
	// do something
}
```

Reply:
Thank you @muraee, but I think what I have is a little simpler :)

Reply:
agreed, I was just having fun trying to figure this out :)